File: MASTER DSA/132 PATTERN/Solution.java
import java.util.*;

class Solution {
    public boolean find132pattern(int[] nums) {
        int n = nums.length;

        Stack<Integer> st = new Stack<>();

        int min[] = new int[n];
        min[0] = nums[0];

        for (int i = 1; i < n; i++)
            min[i] = Math.min(min[i - 1], nums[i]);
        
        for (int i = n - 1; i >= 0; i--) {
            while (st.size() > 0 && st.peek() <= min[i])
                st.pop();
            if (st.size() > 0 && st.peek() < nums[i]) {
                return true;
            }
            st.push(nums[i]);
        }

        return false;
    }
}

File: MASTER DSA/132 PATTERN/Solution.py
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        n = len(nums)

        minTo = []
        minTo.append(nums[0])

        for i in range(1, len(nums)):
            minTo.append(min(nums[i], minTo[i - 1]))

        stack = []

        for i in range(n - 1, 0, -1):
            while len(stack) > 0 and minTo[i] >= stack[-1]:
                stack.pop()

            if len(stack) > 0 and stack[-1] < nums[i]:
                return True

            stack.append(nums[i])

        return False
